### PR TITLE
syntax: subst: fix possible leak of Tm_delayed on compress

### PR DESCRIPTION
**Pull request by mtzguido**
Another issue with `compress`.. from the commit message:

----

It was possible for compress to return a Tm_delayed node when it entered
the Tm_name case and the substitution on the node mapped the variable to
a Tm_delayed node.

The error comes since, at a first glance, `subst'` appears to never
return a Tm_delayed when called on `Tm_type`, `Tm_bvar` and `Tm_name`,
which are the cases on which its called in `push_subst`. However this
assumption is broken when the substitution itself contains a Tm_delayed.

So just recurse in compress. Hopefully the performance impact is
negligible, and this makes it evident that we never return a Tm_delayed.
Also, leave a commented-out check for this property for the future.

-----

I will measure the performance impact since I think this was unrolled for performance reasons.

### CODE REVIEW

**Comment by mtzguido** [comment]
Here's output from benchmarking with a new `--compare` option. It shows the summary on each side and a diff of the data for the top 10 files on each side*.

[*] It also makes sure to show the right-side data for the top 10 files on the left, and viceversa, but they happened to match exactly everywhere here.

Below `guido_tm_delayed` contains the change and `guido_tm_delayed^^^` is just master as of yesterdy at some time.

```
For micro-benchmarks
======================================================================

Summary for guido_tm_delayed^^^
N_benches       total       user     system    mem(mb)
89            27.2602    24.4194    2.78653      10811

Summary for guido_tm_delayed
N_benches       total       user     system    mem(mb)
89            26.6675    24.1238      2.494      10662

Per-file comparison, total time. Left is guido_tm_delayed^^^, right is guido_tm_delayed
1	"ChrisCheck.fst.checked"                	   8.93802	1	"ChrisCheck.fst.checked"                	   9.18464
2	".depend"                               	  0.762769	2	"Unit1.Basic.fst.checked"               	  0.724138
3	"Unit1.Basic.fst.checked"               	   0.72757	3	".depend"                               	  0.711586
4	"Pruning.fst.checked"                   	  0.713613	4	"Pruning.fst.checked"                   	  0.706803
5	"Test.HyperStack.fst.checked"           	  0.656311	5	"Test.HyperStack.fst.checked"           	   0.63851
6	"UnifyMatch.fst.checked"                	  0.594251	6	"UnifyMatch.fst.checked"                	  0.588541
7	"Test.BufferView.fst.checked"           	  0.540605	7	"Test.BufferView.fst.checked"           	  0.529327
8	"StrictUnfolding.fst.checked"           	  0.513309	8	"StrictUnfolding.fst.checked"           	  0.498407
9	"Test.Integers.fst.checked"             	  0.441479	9	"Test.Integers.fst.checked"             	  0.435874
10	"Test.Printf.fst.checked"               	  0.419551	10	"Test.Printf.fst.checked"               	  0.403942

Per-file comparison, memory. Left is guido_tm_delayed^^^, right is guido_tm_delayed
1	"ChrisCheck.fst.checked"                	    694352	1	"ChrisCheck.fst.checked"                	    543300
2	"Pruning.fst.checked"                   	    338660	2	"Pruning.fst.checked"                   	    337960
3	"StrictUnfolding.fst.checked"           	    235056	3	"StrictUnfolding.fst.checked"           	    235052
4	"Unit1.Basic.fst.checked"               	    218656	4	"Unit1.Basic.fst.checked"               	    218820
5	"Test.BufferView.fst.checked"           	    213700	5	"Test.BufferView.fst.checked"           	    214936
6	"UnifyMatch.fst.checked"                	    204992	6	"UnifyMatch.fst.checked"                	    206300
7	"Test.Printf.fst.checked"               	    195364	7	"Test.Printf.fst.checked"               	    194756
8	"IntNormalization.fst.checked"          	    194536	8	"IntNormalization.fst.checked"          	    192508
9	"Test.Integers.fst.checked"             	    182904	9	"Test.HyperStack.fst.checked"           	    181844
10	"Test.HyperStack.fst.checked"           	    181496	10	"Test.Integers.fst.checked"             	    181808

For ulib
======================================================================

Summary for guido_tm_delayed^^^
N_benches       total       user     system    mem(mb)
89            27.2602    24.4194    2.78653      10811

Summary for guido_tm_delayed
N_benches       total       user     system    mem(mb)
89            26.6675    24.1238      2.494      10662

Per-file comparison, total time. Left is guido_tm_delayed^^^, right is guido_tm_delayed
1	"Steel.Channel.Simplex.fst"             	   32.3648	1	"Steel.Channel.Simplex.fst"             	   32.6673
2	"Steel.Semantics.Hoare.MST.fst"         	   21.2479	2	"Steel.Semantics.Hoare.MST.fst"         	   21.5468
3	"FStar.Math.Fermat.fst"                 	   19.2849	3	"FStar.Math.Fermat.fst"                 	   19.6507
4	"Steel.Actions.fst"                     	   14.1809	4	"Steel.Actions.fst"                     	   14.3797
5	"FStar.ModifiesGen.fst"                 	   10.9802	5	"FStar.ModifiesGen.fst"                 	   11.0647
6	"FStar.Tactics.CanonCommSemiring.fst"   	   8.44018	6	"FStar.Tactics.CanonCommSemiring.fst"   	   8.52724
7	"Steel.HigherReference.fst"             	   7.95971	7	"Steel.HigherReference.fst"             	   7.99064
8	"LowStar.Monotonic.Buffer.fst"          	   7.63157	8	"LowStar.Monotonic.Buffer.fst"          	   7.70295
9	"Steel.Effect.Atomic.fst"               	   5.97385	9	"Steel.Effect.Atomic.fst"               	   6.10114
10	"FStar.Tactics.PatternMatching.fst"     	   5.62586	10	"FStar.Tactics.PatternMatching.fst"     	   5.73241

Per-file comparison, memory. Left is guido_tm_delayed^^^, right is guido_tm_delayed
1	"FStar.Math.Fermat.fst"                 	   1493684	1	"FStar.Math.Fermat.fst"                 	   1493648
2	"Steel.Channel.Simplex.fst"             	   1321800	2	"Steel.Channel.Simplex.fst"             	   1321708
3	"FStar.Tactics.PatternMatching.fst"     	   1006924	3	"FStar.Tactics.PatternMatching.fst"     	   1006884
4	"Steel.Semantics.Hoare.MST.fst"         	    787528	4	"Steel.Semantics.Hoare.MST.fst"         	    820348
5	"Steel.HigherReference.fst"             	    707692	5	"Steel.HigherReference.fst"             	    706784
6	"Steel.Effect.Atomic.fst"               	    658740	6	"Steel.Effect.Atomic.fst"               	    658656
7	"FStar.Tactics.CanonCommSemiring.fst"   	    506832	7	"FStar.Tactics.CanonCommSemiring.fst"   	    506772
8	"Steel.Effect.fst"                      	    495156	8	"Steel.Effect.fst"                      	    495092
9	"LowStar.RVector.fst"                   	    484360	9	"LowStar.RVector.fst"                   	    484552
10	"FStar.Math.Euclid.fst"                 	    482176	10	"FStar.Math.Euclid.fst"                 	    482136

For ocaml_extract
======================================================================

Summary for guido_tm_delayed^^^
N_benches       total       user     system    mem(mb)
89            27.2602    24.4194    2.78653      10811

Summary for guido_tm_delayed
N_benches       total       user     system    mem(mb)
89            26.6675    24.1238      2.494      10662

Per-file comparison, total time. Left is guido_tm_delayed^^^, right is guido_tm_delayed
1	"FStar_TypeChecker_Rel.ml"              	   2.81331	1	"FStar_TypeChecker_Rel.ml"              	   2.87482
2	"FStar_SMTEncoding_Encode.ml"           	   2.77528	2	"FStar_SMTEncoding_Encode.ml"           	   2.75571
3	"FStar_TypeChecker_TcTerm.ml"           	    2.5487	3	"FStar_TypeChecker_TcTerm.ml"           	   2.57185
4	"FStar_Tactics_InterpFuns.ml"           	   2.51549	4	"FStar_Tactics_InterpFuns.ml"           	     2.542
5	"FStar_TypeChecker_Normalize.ml"        	   2.28553	5	"FStar_TypeChecker_Normalize.ml"        	   2.30203
6	"FStar_TypeChecker_Util.ml"             	   2.10328	6	"FStar_TypeChecker_Util.ml"             	   2.08298
7	"FStar_ToSyntax_ToSyntax.ml"            	   1.96666	7	"FStar_ToSyntax_ToSyntax.ml"            	   1.94739
8	"FStar_SMTEncoding_EncodeTerm.ml"       	   1.94067	8	"FStar_SMTEncoding_EncodeTerm.ml"       	   1.93504
9	"FStar_Extraction_ML_Term.ml"           	   1.83463	9	"FStar_Extraction_ML_Term.ml"           	   1.85402
10	"FStar_TypeChecker_DMFF.ml"             	   1.66574	10	"FStar_TypeChecker_DMFF.ml"             	   1.67252

Per-file comparison, memory. Left is guido_tm_delayed^^^, right is guido_tm_delayed
1	"FStar_Tactics_InterpFuns.ml"           	    411196	1	"FStar_Tactics_InterpFuns.ml"           	    412292
2	"FStar_Tests_Test.ml"                   	    311468	2	"FStar_Tests_Test.ml"                   	    312344
3	"FStar_Tests_Norm.ml"                   	    298480	3	"FStar_Tests_Norm.ml"                   	    297356
4	"FStar_Tests_Pars.ml"                   	    293612	4	"FStar_Tests_Pars.ml"                   	    293548
5	"FStar_Tests_Unif.ml"                   	    286808	5	"FStar_Tests_Unif.ml"                   	    289308
6	"FStar_Main.ml"                         	    277408	6	"FStar_Main.ml"                         	    279008
7	"FStar_SMTEncoding_Encode.ml"           	    265064	7	"FStar_SMTEncoding_Encode.ml"           	    263732
8	"FStar_Universal.ml"                    	    255784	8	"FStar_Universal.ml"                    	    254968
9	"FStar_TypeChecker_TcTerm.ml"           	    254088	9	"FStar_TypeChecker_TcTerm.ml"           	    254584
10	"FStar_Interactive_Ide.ml"              	    246796	10	"FStar_Interactive_Ide.ml"              	    246884
```
In summary it looks like it just has a small random impact, some files are slower and some improve.
